### DISCUSSION TITLE
Remove the Typescript Extractor?

### DISCUSSION THREAD

**Discussion by nathanforce** [lingui/js-lingui#1047] [Q&A]
What do you think about removing the [Typescript extractor](https://github.com/lingui/js-lingui/blob/main/packages/cli/src/api/extractors/typescript.ts) in favor of always using Babel? Babel supports parsing TS nowadays so keeping both feels redundant.  Also, the current typescript extractor reports incorrect line numbers when creating `origin`. This is because the code that is passed to Babel is first [transpiled by the Typescript API](https://github.com/lingui/js-lingui/blob/26e9df5e1cfe505e4f53e896f8152aa7a75faa59/packages/cli/src/api/extractors/typescript.ts#L22-L31).

**Comment by semoal** [comment]
Yeah was thinking just the same. I'm just trying to find a way of being sure that we don't brake any setup.
I've seen some issues related to the typescript extractor, like the required NODE_ENV=development in lingui extract command sometimes is required to make it work...

I don't want to release a breaking change, so any suggestion it's welcome

**Reply by nathanforce** [reply]
Maybe the best solution is to expose the extractors and allow users to choose. I see a few ways of doing this. Something similar to Jest's `moduleNameMapper` where you map an `extension => extractor`.  Current behavior could be maintained for now with

```
extractors: {
   babel: babel.defaultExtensions,
   typescript: ['.ts', '.tsx']
}
```

Alternatively you could keep the current `{ match, extract }` signature and accept something like: 

```
extractors: ['babel', 'typescript']
```

Both of these options open up the path for supporting custom extractors if we modified the extractor signature to return the extracted messages rather than relying on `babel-plugin-extract-messages` to write the catalog. (IMO this change should happen anyhow as the code is clearer/separated concerns).
